File: fixtures/scale/dart/large/lib/ciphers/vigenere/vigenere_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'vigenere_key.dart';

/// Adds the repeating keyword letter to each plaintext letter.
class VigenereCipher implements Cipher {
  final VigenereKey key;

  VigenereCipher(this.key);

  @override
  String name() => 'vigenere';

  @override
  String encrypt(String plaintext) => _transform(cleanAlphabet(plaintext), true);

  @override
  String decrypt(String ciphertext) => _transform(cleanAlphabet(ciphertext), false);

  String _transform(String text, bool forward) {
    final sb = StringBuffer();
    for (var i = 0; i < text.length; i++) {
      final x = indexOf(text[i]);
      final k = indexOf(key.keyCharAt(i));
      sb.write(charAt(forward ? x + k : x - k));
    }
    return sb.toString();
  }
}

File: fixtures/scale/java/large/src/cryptobench/ciphers/vigenere/VigenereCipher.java
package cryptobench.ciphers.vigenere;

import cryptobench.core.Cipher;
import cryptobench.util.Alphabet;

/** Adds the repeating keyword letter to each plaintext letter. */
public class VigenereCipher implements Cipher {
    private final VigenereKey key;

    public VigenereCipher(VigenereKey key) {
        this.key = key;
    }

    @Override
    public String name() {
        return "vigenere";
    }

    @Override
    public String encrypt(String plaintext) {
        return transform(Alphabet.clean(plaintext), true);
    }

    @Override
    public String decrypt(String ciphertext) {
        return transform(Alphabet.clean(ciphertext), false);
    }

    private String transform(String text, boolean forward) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            int x = Alphabet.indexOf(text.charAt(i));
            int k = Alphabet.indexOf(key.keyCharAt(i));
            sb.append(Alphabet.charAt(forward ? x + k : x - k));
        }
        return sb.toString();
    }
}
